### PR TITLE
CRD Function spec fully implemented 

### LINKED ISSUE #6: Implement function specs (faas-netes port)
Checklist:

Version 1
- [x] replicas
- [x] labels
- [x] env vars

Version 2
- [x] constrains 
- [x] limits
- [x] resources
- [x] secrets
- [x] available replicas

### PR DESCRIPTION
**Pull request by stefanprodan**
Fix #6  This PR implements:

- constrains 
- limits
- resources
- secrets
- available replicas reader
- faas-provider info handler

AMD64 version tested on GKE 1.9.3.
ARMHF version tested on Scaleway with Kubernetes 1.9.5.

Tested with UI, CLI and kubectl: create, update, delete, list, invoke, scale up/down.

When TravisCI is configured it will build and push the image to Docker Hub at functions/faas-o6s ref #5

### CODE REVIEW

**Comment by stefanprodan** [comment]
I've tested once more all the API calls via UI/CLI/kubectl on armhf with faas-o6s as a Gateway sidecar. Everything works as expected.

**Comment by alexellis** [comment]
Why did the design change to a sidecar? I don't remember us discussing that as part of the design.

**Comment by alexellis** [comment]
The commit messages are basically empty on this PR. We're going to need to go through them and pad them out and give:

Why you changed it
How you changed it
And how it was tested

Also I noticed a typo in the commit message for one of the latest commits: "Secretes spec implemented" - should be "Secrets"

Here are some tips for re-writing history - https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

Try not to squash the commits as there is a significant amount of work here - too much to review in a single PR.

File: pkg/server/info.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+	"github.com/openfaas/faas-provider/types"
+)
+
+// makeInfoHandler provides the system/info endpoint
+func makeInfoHandler() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		sha, release := version.GetReleaseInfo()
+		info := types.InfoRequest{
+			Orchestration: "kubernetes",
+			Provider:      "faas-06s",
```
Comment (alexellis) [inline pkg/server/info.go]: Should this be faas-o6s - rather than 06s?

File: pkg/server/info.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/openfaas-incubator/faas-o6s/pkg/version"
+	"github.com/openfaas/faas-provider/types"
+)
+
+// makeInfoHandler provides the system/info endpoint
+func makeInfoHandler() http.HandlerFunc {
```
Comment (alexellis) [inline pkg/server/info.go]: Do we have a test for this? You can copy my one if you need to from my PR to Nic's faas-nomad

File: Dockerfile.armhf [hunk 3]
```diff
@@ -0,0 +1,24 @@
+FROM golang:1.9
```
Comment (alexellis) [inline Dockerfile.armhf]: Probably wants pinning to a minor version

File: README.md [hunk 4]
```diff
@@ -61,7 +89,7 @@ Start OpenFaaS controller (assumes you have a working kubeconfig on the machine)
 
 ```bash
 $ go build \
-  && ./faas-o6s -kubeconfig=$HOME/.kube/config -logtostderr=true
+  && ./faas-o6s -kubeconfig=$HOME/.kube/config -logtostderr=true -v=4
```
Comment (alexellis) [inline README.md]: What does `-v=4` mean here? (why did it change?)

File: README.md [hunk 5]
```diff
@@ -50,6 +45,39 @@ kubectl -n openfaas-fn get functions
...
+  constraints:
+    - "cloud.google.com/gke-nodepool=default-pool"
+```
+
+Apply the function and check the deployment specs with:
+
+```bash
+kubectl -n openfaas-fn describe deployment gofast
+```
+
```
Comment (alexellis) [inline README.md]: Can we add `### Run locally with Docker`?

File: README.md [hunk 6]
```diff
@@ -12,35 +12,30 @@ cd faas-netes
...
 # CRD
 kubectl apply -f artifacts/o6s-crd.yaml
 # RBAC
 kubectl apply -f artifacts/o6s-rbac.yaml
-# Service
-kubectl apply -f artifacts/o6s-svc.yaml
-# Deployment
-kubectl apply -f artifacts/o6s-dep.yaml
+# Deployment (use o6s-armhf.yaml for faas-netes/yaml_armhf)
+kubectl apply -f artifacts/o6s-amd64.yaml
```
Comment (alexellis) [inline README.md]: Can the folders be separated out like faas-netes / faas for arm / arm64 etc?

**Comment by alexellis** [comment]
Can we switch to a non-root user too like in faas-netes please?

**Comment by alexellis** [comment]
Commit messages much more useful now. Thank you.

**Pull request merged**